### PR TITLE
fix(diagnostics): reject invalid literal references

### PR DESCRIPTION
**Pull request by paulfalgout**
Related #134

## What

- Add stable MN0018 diagnostics for literal @ui references that are unknown, inherited-only, or not declared by the owning View or Behavior.
- Add stable MN0019 diagnostics for invalid handler values and missing or non-callable named handlers, and preflight complete DOM maps before any listener is delegated.
- Merge Behavior UI with host View UI before initial delegation, document the contracts, and update generated distribution artifacts.

## Why

Literal reference mistakes currently survive until incidental selector or invocation failures, and partially valid event maps can delegate listeners before a later entry throws. These failure modes are hard for agents to diagnose and can leave partially configured Views. The new diagnostics identify the invalid definition at construction/delegation time while preserving direct functions and inherited callable named handlers.

## Scope

- Affects View and Behavior UI resolution, DOM event/trigger delegation, the shared method resolver, diagnostic catalog, current docs, migration notes, tests, and the six generated main distribution files/maps.
- Does not change Region-name diagnostics, Radio/reset diagnostics, post-destroy lifecycle semantics, package exports, dependencies, or GitHub settings.
- This is a bounded seed conversion related to #134; it does not complete the remaining diagnostic conversion inventory.

## Deployment Impact

No application or documentation-site deployment is performed by this PR. Package consumers receive the behavior correction only in a later explicitly authorized Marionette release. No website, npm package, tag, or release is published here.

## Runtime cost

- Exact-base Brotli-11 total: 49,822 B to 50,655 B, +833 B (+1.67%).
- Current cumulative size remains below the base-owned ceiling: 50.66 kB / 51.98 kB.
- No new runtime artifact, public subpath, module graph, dependency, eager allocation, or retained-resource shape is added.
- The four main artifacts cross the 1% exact-base threshold and therefore require the canonical exact-head growth approval; no budget amendment is required.

## Testing

- `npx vitest run test/unit/behavior-ui-contract.spec.js test/unit/common/normalize-methods.spec.js test/unit/mixins/ui.spec.js test/unit/view.ui-bindings.spec.js test/unit/view.ui-event-and-triggers.spec.js --coverage.enabled=false --reporter=dot` - 5 files, 45 tests passed.
- `npm test -- --coverage.enabled=false --reporter=dot` - 69 files, 1,184 tests passed.
- `npm run test:performance` - 89 tests passed.
- `npm run lint:ci`
- `npm run check:diagnostics` - 19 entries validated.
- `npm run docs:check` - 31 pages, 32 HTML files, 304 internal links.
- `npm run size`
- `npm run check:dist`
- `npm run test:dist`
- `npm run test:source`
- `npm run test:fixtures`
- `git diff --check`

The rebuilt distribution artifacts were byte-stable after rebasing onto master.

---
## Summary by cubic
Adds stable diagnostics for invalid `@ui` references and event handler values so configuration mistakes fail at construction time instead of surfacing as incidental selector or invocation failures. Previously, missing `@ui` keys normalized to `undefined` and non-callable handlers were silently dropped; now `MarionetteError` is thrown with codes `MN0018` and `MN0019`, and the complete DOM event map is preflighted before any listener is delegated. Behavior `ui` maps are merged with the host View's before initial delegation, with host selectors winning on conflicts. `MN0019` also applies to all normalized handler maps, such as model, collection, Radio, request, and child-view bindings.

**Migration**
- Update any Views or Behaviors that reference `@ui` keys that aren't non-empty own entries, or that use handler values that aren't functions or names of callable methods; those will now throw `MarionetteError` with codes `MN0018`/`MN0019`.

<sup>Written for commit 73716eae81d3052c075750bbb65cdbe1f729f13b. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/marionettejs/marionette/pull/208?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added clear runtime diagnostics for missing UI references (`MN0018`) and invalid event handlers (`MN0019`).
  * Added validation for UI references and event, request, and trigger handlers before binding or delegation.
  * Improved Behavior UI resolution so host selectors and overrides are available immediately.

* **Documentation**
  * Updated usage and migration guidance for UI reference and handler validation, including remediation details.

* **Tests**
  * Added coverage for validation errors, inherited and empty UI keys, handler edge cases, and Behavior UI bindings.

### CODE REVIEW

**Comment by paulfalgout** [comment]
```json
{
  "schemaVersion": 1,
  "headSha": "73716eae81d3052c075750bbb65cdbe1f729f13b",
  "issueUrl": "https://github.com/marionettejs/marionette/issues/127",
  "approvedPaths": [
    "dist/marionette.cjs",
    "dist/marionette.js",
    "dist/marionette.min.js",
    "dist/marionette.umd.js"
  ],
  "evidenceUrls": [
    "https://github.com/marionettejs/marionette/issues/127#issuecomment-5457438341"
  ]
}
```

**Comment by paulfalgout** [comment]
Reviewed at `cfc55f77` (the SHA I fetched). **Not approving yet** — CI is red for two distinct reasons, and one of them points at a gap in the diagnostic itself rather than at a missing test.

**Bundle measurement independently confirmed.** I measured brotli-11 across the eight contract artifacts from the committed `dist/` and reproduce your figure exactly:

```
marionette.js      12866 -> 13068  +202 (1.57%)
marionette.cjs     13010 -> 13223  +213 (1.64%)
marionette.umd.js  13285 -> 13511  +226 (1.70%)
marionette.min.js  10067 -> 10232  +165 (1.64%)
AGGREGATE          49822 -> 50628  +806 (1.62%)
```

All four cross 1% individually, so the exact-head growth approval is genuinely required and `Bundle size` failing without an approval comment is the gate behaving correctly. Worth stating the cost in consumer terms as well: **165 min-build bytes**, leaving roughly **276 min-build bytes** under the ceiling. This one PR spends about 37% of the remaining core-code budget, and it describes itself as "a bounded seed conversion… it does not complete the remaining diagnostic conversion inventory." The amendment protocol from #207 landed one PR before it's actually needed.

**`Node 24` fails on the 100% coverage gate:** `Coverage for branches (99.85%) does not meet global threshold (100%)`, isolated to `mixins/view-events.js` line 74 (97.5% statements, 95.23% branches).

That line is `if (!handler) { return; }`, and I traced when it's now reachable. `resolveMethod` throws MN0019 only when `isString(methodName) && !isFunction(resolvedMethod)`. So a falsy return survives exactly when the handler is **neither a function nor a string** — `events: { 'click .x': undefined }`, `null`, or a number — where `context[method]` is falsy and the string guard never fires.

So the branch isn't dead, but it documents an inconsistency: a *typo'd string* handler now raises MN0019 at delegation time, while a `null` or `undefined` handler is still silently skipped. Given the stated goal — "literal reference mistakes currently survive until incidental selector or invocation failures… the new diagnostics identify the invalid definition at construction/delegation time" — an `undefined` handler is at least as clear a mistake as a misspelled method name, and arguably easier to introduce (a stray comma, a renamed import, a conditional that returned nothing).

I'd resolve it as a behavior question before a coverage one. Either extend MN0019 to any non-function handler that fails to resolve, which removes the branch entirely and makes the diagnostic's contract "a handler must resolve to a function, whatever you passed"; or deliberately keep the silent skip for non-strings and add the test that pins it, with a note saying why strings and non-strings differ. The first reads more consistent with the rest of the diagnostic work; the second is defensible if legacy callers rely on falsy entries being ignored.

`Build verified tarball` is also failing — I haven't traced whether that's the same coverage gate or independent; worth confirming it isn't a third issue.

Happy to re-review once CI is green.

<sub>Reviewed with Claude Code.</sub>

**Comment by paulfalgout** [comment]
Exact-head Claude re-review requested for `4a9d0c52a33a7230fbc4473995577abc19867ace`. This implements the preferred canonical resolution from the prior review: MN0019 now rejects every handler value that is neither a function nor a string resolving to a callable own or inherited method. The silent numeric/null/undefined skip path and its delegation branch are removed; catalog/docs/migration guidance now state the complete rule. Exact-base Brotli growth decreased from +806 B/+1.62% to +788 B/+1.58% (50,610 B / 51,975 B), with no graph, artifact, allocation, or retention addition. Local validation on this head: coverage 100% (1,438 statements, 702 branches, 375 functions, 1,353 lines), 1,181 unit tests, 19 agent benchmark contract tests, 89 performance tests, lint, diagnostics, docs, size, dist/source/package fixtures, and diff check. The old growth approval is intentionally stale; approval refresh infrastructure is being repaired separately before a final exact-head approval.

**Comment by paulfalgout** [comment]
Re-reviewed at `4a9d0c52` (the SHA I fetched). **Approve**, pending only the growth-approval comment.

You took the widening option, and it's carried through completely rather than patched at the call site. `resolveMethod` now looks up `context[methodName]` only for strings and throws for any non-function result, the `if (!handler) { return; }` branch in `view-events.js` is *deleted* rather than covered — correct, since it's genuinely unreachable now — and `normalizeMethods` drops its `if (method)` guard for the same reason. Renaming the catalog slug from `string-handler-not-found` to `handler-not-callable` is the part I'd have asked for if it were missing: the slug and `docsAnchor` are the agent-facing identity, and leaving "string" in the name of a diagnostic that no longer only concerns strings would have been a small lie with a long half-life. The migration row was widened to match.

**Coverage is green** — `Node 24` and `Build verified tarball` both pass, so that was a single root cause rather than the third issue I flagged as unverified.

**Re-measured after the change:**

```
AGGREGATE  49822 -> 50610   +788 (+1.58%)
min-build  +144 bytes
headroom   1365 aggregate  ~=  249 min-build bytes
```

Slightly cheaper than `cfc55f77`'s +806 — removing the two now-dead branches gave back 18 bytes. Still over 1% on the four main artifacts, so the exact-head approval is still required.

**One incidental finding from the failing job's log, which is the clearest evidence yet for something I've been reporting from file comparisons.** Line 467 of the `Bundle size` run reads:

```
node "${authority_script}" --config "${base_contract}" --json > …/bundle-size-authority.json
```

That's the **pre-#194** command form. The activation wiring on master passes `--config config/performance.json` with `--candidate-contract`. So this run is executing the workflow pinned at `50808912`, confirming in a live job that the ruleset pin has not moved. It doesn't affect this PR — the 1% growth approval predates #194 and is present in the pinned copy, and #208 doesn't touch `config/performance.json` — but it means new-subpath enforcement is still inert, and the next PR that adds a subpath will discover that the hard way.

<sub>Reviewed with Claude Code.</sub>

File: mixins/ui.js [hunk 1]
```diff
@@ -12,12 +13,23 @@ const normalizeUIKeys = function(hash, ui) {
...
 const uiRegEx = /@ui\.[a-zA-Z-_$0-9]*/g;
+const hasOwnProperty = Object.prototype.hasOwnProperty;
 
 // utility method for parsing @ui. syntax strings
 // into associated selector
 const normalizeUIString = function(uiString, ui) {
   return uiString.replace(uiRegEx, (r) => {
-    return ui[r.slice(4)];
+    const name = r.slice(4);
+    const selector = ui && ui[name];
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply mixins/ui.js]: Fixed in `802c2282`: ownership is now checked before selector access. A regression test uses an inherited throwing getter and proves MN0018 is returned without invoking the accessor.

File: modules/behavior.js [hunk 2]
```diff
@@ -44,6 +43,7 @@ const Behavior = function(options, view) {
   // between multiple views, while letting a view override
   // a selector under an UI key.
   this.ui = _extend({}, result(this, 'ui'), result(view, 'ui'));
+  this.setElement();
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/behavior.js:47]: Fixed in `802c2282`: the constructor assigns the host element before evaluating functional Behavior `ui`, while retaining one post-merge `setElement()` delegation pass. The regression test proves functional UI sees the host element, runs once, and delegates its selector.

File: modules/common/normalize-methods.js [hunk 3]
```diff
@@ -1,22 +1,35 @@
...
+const resolveMethod = function(context, method, name) {
+  if (isFunction(method)) { return method; }
+
+  const methodName = method;
+  const resolvedMethod = isString(methodName) ? context[methodName] : undefined;
+
+  if (!isFunction(resolvedMethod)) {
+    throw new MarionetteError({
+      code: 'MN0019',
+      message: `The handler "${methodName}" for "${name}" must resolve to a function.`
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/common/normalize-methods.js]: Fixed in `802c2282`: MN0019 formats invalid handler references with `String(methodName)`. A Symbol regression test now proves the stable Marionette diagnostic replaces the native interpolation TypeError.

**Comment by paulfalgout** [comment]
Re-reviewed at `802c2282` (the SHA I fetched). **Approve**, still pending only the growth-approval comment.

**`String(methodName)` is the widening biting back, and catching it here was sharp.** Once MN0019 throws for *any* non-function handler rather than only unresolvable strings, a Symbol handler reaches that template literal — and interpolating a Symbol throws rather than stringifying:

```
`${Symbol("handler")}`  ->  TypeError: Cannot convert a Symbol value to a string
String(Symbol("handler")) ->  Symbol(handler)
```

So without this, widening the diagnostic would have turned a silent skip into the *wrong* failure for Symbol handlers: a raw `TypeError` from inside the error constructor instead of the MN0019 the contract promises. That's strictly worse than either the old or intended behavior, and it's only reachable because the previous commit removed the non-string escape hatch.

**`this.el = view.el` before the UI merge** — I checked whether this introduces a stale snapshot, since the Behavior contract documented in #198 says the `el` is the host's *current* element. It doesn't: `setElement()` still runs later in construction, and `mixins/behaviors.js:62` `_setBehaviorElements()` maps `setElement()` across every behavior when the host's element changes. The new assignment only opens the window earlier so a functional `ui` can read `this.el` while being evaluated. The test asserting `this.el === this.view.el` *inside* the `ui` function pins the property that actually matters rather than just the return value.

**The `mixins/ui.js` refactor** reads as equivalent, with one small improvement worth noting: the old form evaluated `ui[name]` before the own-property check could reject it, so a prototype-chain getter would run before the throw. Computing `hasSelector` first means the lookup happens only when the own property exists.

Bundle impact unchanged in character; I'll re-measure if the head moves again before the approval lands.

<sub>Reviewed with Claude Code.</sub>

**Comment by paulfalgout** [comment]
Exact-head re-review requested for `809c04eb54ae9e08d47494f0c09b19d0b41e5ddd`. This is the previously reviewed three-commit diagnostic change rebased unchanged onto merged #209; the runtime/catalog/docs diff remains the same, with no conflicts. Local focused tests, full coverage, diagnostics, performance, lint, source/dist/fixture suites, docs, size, release-profile/promotion, artifact equality, diff, and clean-status checks all pass. Please review this exact head; growth approval will be recorded after exact-base CI reports the final bytes.

**Comment by paulfalgout** [comment]
Re-reviewed at `809c04eb` (the SHA I fetched). **Approve.** `Bundle size` is green — the growth-approval chain completed its first real end-to-end run.

Worth recording what that exercised, since it spans four merged PRs: an approval comment binding the exact head, the refresh workflow locating the organization-required run through the `required_workflows/` lookup from #209, and the rerun accepting growth above the 1% threshold. Prior to #209 that lookup returned nothing and the refresh could never fire, so this is the first time the whole path has run against a live payload rather than fixtures.

I verified the rebase onto `3b0bdd35` was content-preserving rather than assuming it: `git diff 802c2282 809c04eb -- dist/` is empty, and excluding `dist/` plus the two files #209 brought in, the source diff is empty too.

**Final measurement at this head**, against the new base:

```
AGGREGATE  49822 -> 50655   +833 (+1.67%)
min-build  +146 bytes
headroom   1320 aggregate  ~=  231 min-build bytes
```

Slightly above the +788 I reported at `4a9d0c52` — that difference is `802c2282`'s content (the `this.el` assignment, the `String()` Symbol fix, the `ui.js` refactor and their tests), not rebuild drift. I checked, because a byte change from unchanged source would have contradicted the roadmap's reproducibility expectation and been worth raising on its own.

Nothing open from me.

<sub>Reviewed with Claude Code.</sub>

File: modules/common/normalize-methods.js [hunk 4]
```diff
@@ -1,22 +1,35 @@
...
+const resolveMethod = function(context, method, name) {
+  if (isFunction(method)) { return method; }
+
+  const methodName = method;
+  const resolvedMethod = isString(methodName) ? context[methodName] : undefined;
+
+  if (!isFunction(resolvedMethod)) {
+    throw new MarionetteError({
+      code: 'MN0019',
+      message: `The handler "${String(methodName)}" for "${name}" must resolve to a function.`
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/common/normalize-methods.js]: Fixed in `6698b37170b4ec680080806460fdcf7c33498412`. Invalid handlers now use a `<unprintable>` fallback if string coercion throws, so null-prototype and other hostile values still produce MN0019. Added a regression test using `Object.create(null)`; focused tests, full coverage, lint, source/dist/fixtures, performance, and artifact checks pass.

File: mixins/ui.js [hunk 5]
```diff
@@ -12,12 +13,24 @@ const normalizeUIKeys = function(hash, ui) {
...
+    const selector = hasSelector ? ui[name] : undefined;
+
+    if (!hasSelector || typeof selector === 'undefined') {
+      throw new MarionetteError({
+        code: 'MN0018',
+        message: `The ui reference "${name}" is not defined.`
+      });
+    }
+
+    return selector;
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply mixins/ui.js:48]: Fixed in `6698b37170b4ec680080806460fdcf7c33498412`. Literal `@ui` expansion now requires an own declared string selector and reports MN0018 otherwise. Direct `_bindUIElements` still accepts non-string selectors; a regression test pins that distinction. Focused tests, full coverage, lint, source/dist/fixtures, performance, and artifact checks pass.

**Comment by paulfalgout** [comment]
Exact-head re-review requested for `6698b37170b4ec680080806460fdcf7c33498412`. Both Cubic findings are fixed and their threads resolved: unprintable invalid handlers retain MN0019, and literal `@ui` expansion requires a declared string while direct UI binding keeps accepting non-string selectors. Regression tests cover both distinctions. Focused tests, full 100% coverage, diagnostics, performance, lint, source/dist/fixture suites, docs, size, release-profile/promotion, artifact reproducibility, diff, and clean-status checks pass. Growth evidence/approval will be refreshed from exact-base CI for this head.

**Comment by paulfalgout** [comment]
Re-reviewed at `6698b371` (the SHA I fetched). **Approve.** Both changes continue the thread the widening opened, and the first one closes a hole the previous fix left.

**`String()` can itself throw**, which I confirmed rather than assumed:

```
String(Object.create(null))  ->  TypeError: Cannot convert object to primitive value
```

A null-prototype object has no `Symbol.toPrimitive`, `toString`, or `valueOf`, so the last commit's `String(methodName)` — which correctly fixed Symbols — still crashed inside the error constructor for that input. The result would have been a raw `TypeError` escaping in place of MN0019, exactly the failure mode the Symbol fix addressed, just one input class further out. Wrapping it with an `<unprintable>` fallback makes the guarantee unconditional: whatever you pass, you get MN0019. A diagnostic contract that an exotic value can knock out isn't a contract, so this is the right place to be thorough rather than clever.

**`typeof selector === 'undefined'` → `!isString(selector)`** extends MN0018 the same way MN0019 was extended: a `ui` entry holding a number, object, `null`, or function is now rejected at normalization instead of being handed to the selector engine to fail incidentally later. `isString` is already imported at the top of `mixins/ui.js`, so no new dependency. The catalog remediation was updated to "define the referenced key **as a string selector**", which keeps the agent-facing text aligned with the widened check — the same discipline as renaming the MN0019 slug when its scope grew.

Worth noting the pattern across these three commits: each widening surfaced the next input class that could defeat the diagnostic — non-string handlers, then Symbols, then values with no primitive coercion. That's a reasonable way to converge, and the tests pin each step by asserting the *message text*, so a regression would show up as a wrong string rather than a silently different error type.

Bundle impact: I'll re-measure if this is the head that merges, since the last figure I published was for `809c04eb`.

<sub>Reviewed with Claude Code.</sub>

File: modules/common/normalize-methods.js [hunk 6]
```diff
@@ -1,22 +1,42 @@
...
+
+  if (!isFunction(resolvedMethod)) {
+    let methodLabel = '<unprintable>';
+    try {
+      methodLabel = String(methodName);
+    } catch {
+      // Preserve the stable fallback for values without string coercion.
+    }
+
+    throw new MarionetteError({
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply modules/common/normalize-methods.js:23]: Confirmed intentional at `ef3396a25630df550781b75858a90eae53a99a27`; no code narrowing. MN0019 is the canonical contract for every supplied handler map, including selective `unbindEvents`/`unbindRequests`, as already stated in `docs/common.md`. Silently skipping an unresolved selective unbind would falsely imply the requested handler was removed. Callers that need unconditional cleanup can omit the map, which uses `stopListening(entity)` or context-wide `stopReplying`; the migration row now names selective unbinding and that alternative explicitly. This keeps one strict public normalization path without a lenient teardown fallback.

File: mixins/ui.js [hunk 7]
```diff
@@ -12,12 +13,24 @@ const normalizeUIKeys = function(hash, ui) {
...
   return uiString.replace(uiRegEx, (r) => {
-    return ui[r.slice(4)];
+    const name = r.slice(4);
+    const hasSelector = ui && hasOwnProperty.call(ui, name);
+    const selector = hasSelector ? ui[name] : undefined;
+
+    if (!hasSelector || !isString(selector)) {
+      throw new MarionetteError({
+        code: 'MN0018',
+        message: `The ui reference "${name}" is not defined.`
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply mixins/ui.js]: Fixed in `ef3396a25630df550781b75858a90eae53a99a27`. Missing own keys retain the “not defined” MN0018 message; declared non-string values now receive “must be a string selector.” The regression test pins the distinct message while direct `_bindUIElements` continues accepting non-string selectors. Focused/full coverage, lint, docs, diagnostics, performance, source/dist/fixtures, reproducibility, and size checks pass.

**Comment by paulfalgout** [comment]
Exact-head re-review requested for `ef3396a25630df550781b75858a90eae53a99a27`. The new MN0018 wording finding is fixed. The MN0019 blast radius is intentionally retained and now explicit in migration guidance: selective bind/unbind maps are strict configuration contracts, while omitting the map remains the unconditional cleanup path. Both Cubic threads are answered and resolved. Focused tests, full 100% coverage, diagnostics, performance, lint, source/dist/fixture suites, docs, size, release-profile/promotion, artifact reproducibility, diff, and clean-status checks pass. Growth evidence/approval will be refreshed from exact-base CI for this head.

**Comment by paulfalgout** [comment]
Re-reviewed at `ef3396a2` (the SHA I fetched). **Approve.**

**Splitting the MN0018 message is the substantive change here.** With the type check folded into the same branch, `ui: { invalid: 1 }` reported *"The ui reference \"invalid\" is not defined"* — which is actively misleading, since the key is right there and only its type is wrong. Someone reading that goes looking for a missing declaration. *"must be a string selector"* points at the actual defect.

Keeping both cases under MN0018 rather than minting a new code is the right call. The code is the machine-facing identifier agents branch on, and both cases have the same remediation shape — fix the `ui` entry. A separate code would have churned the catalog to distinguish a sub-case that doesn't change what you do about it.

**I traced the new migration claim** that MN0019 now applies to "selective unbinding": `unbindEvents(entity, bindings)` → `normalizeBindings` → `normalizeMethods` → `resolveMethod`, so a non-resolving handler in an unbind map throws. And the documented escape — "omit the map when unbinding everything" — is exactly the `if (!bindings) { this.stopListening(entity); return this; }` short-circuit. Accurate on both halves.

For the record, since I've been strict about this on other pull requests: no test in this branch exercises MN0019 through an unbind path — `git grep -l MN0019` at this head returns three files, none mentioning `unbind`. I'm not raising it, because the backing is mechanism-level one layer down, with `unbindEvents` calling straight into a tested `resolveMethod` and no branch in between that could bypass it. That's the same standard I accepted on #198 for the `dom:refresh` claim, and it would be inconsistent to demand more here.

Worth noting this row now documents a real behavior change for consumers who previously passed sloppy unbind maps and had bad entries silently dropped. The migration table is the right place for it, and it's phrased with the escape hatch, which is what someone hitting it will need.

<sub>Reviewed with Claude Code.</sub>

File: mixins/ui.js [hunk 8]
```diff
@@ -12,12 +13,31 @@ const normalizeUIKeys = function(hash, ui) {
...
   return uiString.replace(uiRegEx, (r) => {
-    return ui[r.slice(4)];
+    const name = r.slice(4);
+    const hasSelector = ui && hasOwnProperty.call(ui, name);
+    const selector = hasSelector ? ui[name] : undefined;
+
+    if (!hasSelector) {
+      throw new MarionetteError({
+        code: 'MN0018',
+        message: `The ui reference "${name}" is not defined.`
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply mixins/ui.js]: Fixed in `eaf775e367aa35b5f6c7974d31235bea96b902a5`. Missing and inherited-only references now say the key “must be declared as an own ui key,” while own non-string values retain the distinct “must be a string selector” message. Focused tests, full 100% coverage, lint, docs, diagnostics, performance, source/dist/fixtures, reproducibility, and size checks pass.

**Comment by paulfalgout** [comment]
Exact-head re-review requested for `eaf775e367aa35b5f6c7974d31235bea96b902a5`. The remaining MN0018 wording finding is fixed: missing/inherited-only references now require declaration as an own ui key, distinct from own non-string values requiring a string selector. The Cubic thread is answered and resolved. Focused tests, full 100% coverage, diagnostics, performance, lint, source/dist/fixture suites, docs, size, artifact reproducibility, diff, and clean-status checks pass. Growth evidence/approval will be refreshed from exact-base CI for this head.

**Comment by paulfalgout** [comment]
Re-reviewed at `eaf775e3` (the SHA I fetched). **Approve.** Message-only, and it fixes the same misdirection the last head fixed for the type case.

*"is not defined"* was wrong in a second way I hadn't called out: the inherited-key case. Your own test at `ui.spec.js` sets `view.ui = Object.create(prototype)` and expects the throw — but in that scenario `view.ui.danger` resolves perfectly well in a console, so a developer reading "is not defined" would conclude the diagnostic was broken rather than that own-ness is the requirement. *"must be declared as an own ui key"* names the rule being enforced.

The two MN0018 variants now read as a matched pair, both stating a requirement instead of a negation:

- `must be declared as an own ui key` — absent, or reachable only through the prototype
- `must be a string selector` — present, wrong type

Also worth noting `expect(inheritedGetter).not.to.have.been.called` in that same test pins the prototype-getter behavior I flagged as a minor gain two heads ago, when `hasSelector` was hoisted above the lookup. I'd assumed that was incidental; it's actually asserted, so a regression that reintroduced the eager read would fail rather than pass quietly.

Nothing open from me.

<sub>Reviewed with Claude Code.</sub>

File: test/unit/view.ui-event-and-triggers.spec.js [hunk 9]
```diff
@@ -103,18 +103,44 @@ describe('view ui event trigger configuration', function() {
...
-      view.render();
-      view.el.querySelector('.foo').click();
-    }).to.not.throw();
+    expect(() => new View())
+      .to.throw('The handler "missingHandler" for "click .foo" must resolve to a function.')
+      .with.property('code', 'MN0019');
+  });
+
+  it('rejects non-string event handlers', function() {
+    const delegate = this.sinon.spy(Marionette.View.prototype.EventDelegator, 'delegate');
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/view.ui-event-and-triggers.spec.js]: Fixed in c54973f. The event preflight tests now inject a per-subclass delegator through View.setEventDelegator, and the direct UI binding test uses a real View through public bindUIElements. Focused tests pass 23/23; full coverage remains 100%.

File: mixins/ui.js [hunk 10]
```diff
@@ -12,12 +13,31 @@ const normalizeUIKeys = function(hash, ui) {
...
   return uiString.replace(uiRegEx, (r) => {
-    return ui[r.slice(4)];
+    const name = r.slice(4);
+    const hasSelector = ui && hasOwnProperty.call(ui, name);
+    const selector = hasSelector ? ui[name] : undefined;
+
+    if (!hasSelector) {
+      throw new MarionetteError({
+        code: 'MN0018',
+        message: `The ui reference "${name}" must be declared as an own ui key.`
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply mixins/ui.js:37]: Fixed in c54973f. Bare @ui. now throws MN0018 with an explicit key-name diagnostic before lookup, even if an own empty-string UI key exists. The focused regression test and docs/catalog remediation distinguish an empty key name from a declared empty selector.

File: test/unit/view.ui-event-and-triggers.spec.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (paulfalgout) [reply test/unit/view.ui-event-and-triggers.spec.js]: Fixed in c54973f. The tests now configure the derived View with setEventDelegator and exercise direct UI binding through a real View and public bindUIElements. Focused tests pass 23/23; full coverage remains 100%.

**Comment by paulfalgout** [comment]
Exact-head Claude re-review requested for c54973f341944cbbdda0f1bd394c73e14837f6ae. This addresses all final review findings: bare @ui. now receives an explicit MN0018 key-name diagnostic even when an own empty-string key exists; event preflight tests inject a per-derived-View delegator through setEventDelegator; and direct UI binding uses a real View through public bindUIElements. All three threads are replied and resolved. Exact-base cumulative Brotli growth is +1,100 B (+2.21%), total 50,922 B / 51,975 B; the incremental final-review fix is +56 B (+0.11%) over eaf775e, with no graph, allocation, or retention change. Local validation on this exact head: 1,188 unit tests and 100% coverage, 19 agent benchmark tests, 89 performance tests, focused 23/23, lint, diagnostics, docs, source, dist, package fixtures, reproducibility, size, and diff checks. Exact evidence and approval are refreshed.

**Comment by paulfalgout** [comment]
Re-reviewed at `c54973f3` (the SHA I fetched). **Approve.**

**The empty-name guard has the fixture it needs to mean something.** `@ui.` yields `name === ''`, and before this a declared `ui['']` key would have satisfied `hasOwnProperty` and resolved the reference — so a stray trailing dot silently bound to an empty-named entry. The test sets `view.ui[''] = '.empty'` *before* asserting the throw, which is the detail that makes it a real test: without that line it would pass through the existing `hasSelector` rejection and prove nothing about the new check.

**The non-string asymmetry is now pinned, and I think that's worth having explicit.** After the previous head, `@ui.x` rejects a non-string selector with MN0018, but `bindUIElements()` still preserves one and hands it to `$` — which the reworked `preserves non-string selectors when binding ui directly` case now covers. That reads like an inconsistency until you see why: `@ui.` performs string substitution into an event key, where a non-string can't work, whereas direct binding just passes the value through to the DOM API. Pinning both halves means nobody later "fixes" one into agreement with the other.

**On the `setEventDelegator` refactor** — I traced this rather than assuming it was a correctness fix, and it isn't one. `mixins/view-events.js:88` calls `this.EventDelegator.delegate(...)`, which resolves through the instance to the prototype, so the old `sinon.spy(View.prototype.EventDelegator, 'delegate')` was genuinely observing the object under test. What the change buys is that the per-class delegator makes *which* object is observed unambiguous, and stops the test mutating shared prototype state. That still matters for the load-bearing assertion — `expect(delegate).not.to.have.been.called` in the preflight case — because an assertion that nothing was delegated is only as trustworthy as the certainty that you were watching the right delegator.

Catalog remediation and `dom.interactions.md` both track the non-empty-name rule, so the machine-facing text and the prose stay in step with the check.

<sub>Reviewed with Claude Code.</sub>

File: config/diagnostics/catalog.json [hunk 11]
```diff
@@ -291,6 +291,45 @@
         "runtime"
       ],
       "benchmarkCategory": "communication"
+    },
+    {
+      "code": "MN0018",
+      "slug": "ui-reference-not-found",
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply config/diagnostics/catalog.json]: Fixed in 73716ea. MN0018 now uses the stable slug ui-reference-invalid, accurately covering empty names, missing or inherited keys, and declared non-string selectors without splitting one normalization invariant across multiple codes.

**Comment by paulfalgout** [comment]
Exact-head Claude re-review requested for 73716eae81d3052c075750bbb65cdbe1f729f13b. This metadata-only amendment resolves the final Cubic finding by widening MN0018 stable slug to ui-reference-invalid; runtime and dist are byte-identical to approved c54973f. The thread is replied and resolved, exact evidence and growth approval are refreshed, and exact-head diagnostics, docs, lint, reproducible dist, and diff checks pass.

**Comment by paulfalgout** [comment]
Re-reviewed at `73716eae` (the SHA I fetched). **Approve.** One line, and it's the same correction you applied to MN0019 when its scope grew.

MN0018 now covers three distinct rejections — a missing key name from `@ui.`, an absent or inherited-only key, and a declared key holding a non-string value. `ui-reference-not-found` described only the middle one, so `ui-reference-invalid` is the accurate umbrella. That matters more for a slug than for prose because it feeds `docsAnchor: /errors/MN0018/` and is the identifier an agent or a docs page keys on; a name that describes a subset of what the code fires for is the kind of thing that quietly misroutes someone diagnosing the other two cases.

I checked the rename didn't strand a reference: `git grep ui-reference-not-found` at this head returns nothing, and the new slug appears only in the catalog entry itself. The validator enforces slug uniqueness (`addDuplicateErrors(diagnostics, 'slug', …)`) rather than any ordering tied to the name, so nothing else needed to move with it.

Nothing open from me.

<sub>Reviewed with Claude Code.</sub>
